example.js

```diff
@@ -1,4 +1,4 @@
-const balanced = require('./')
+const balanced = require('./dist/balanced-match.js')
```

Comment:
node comes with native support in recent versions, I wonder if we need to add a bundler, or can just bump the major and make it an ESM only module, as @sindresorhus has recently done with many modules (eg https://github.com/sindresorhus/p-queue/commit/8c7325a4c59c4ecd1edbc196a36450655821776f)

Reply:
That's certainly an option and sindresorhus has made a case for doing so [on his blog](https://blog.sindresorhus.com/get-ready-for-esm-aa53530b3f77), but doing blindly runs the risk of benefitting library authors at the expense of library consumers.

A few issues I think are worth considering:

1. The package you linked to ([p-queue](https://github.com/sindresorhus/p-queue)) is better positioned to go "ESM only" because the library requires `Promise` support. Practically speaking, this means the library will almost certainly be used in environments that support ESM modules.

1. Is this library likely to run in a legacy environment? If so, then I believe the library should offer legacy-compatible code.

   The alternative is to force library consumers to wrestle with webpack, rollup, parcel, etc. Even if these tools are already being used in a project, they have to be (re)configured to transpile some-but-not-all dependencies in `/node_modules` to void issues. This is the issue I ran into when updating `balance-match` to v2 in [css-vars-ponyfill](https://github.com/jhildenbiddle/css-vars-ponyfill): I was already using babel, but it was configured to ignore files in `/node_modules`. As a result, rollup+babel created an ES5 bundle that contained properly transpired code (my source) and untranspiled ES6 code (balance-match v2). See https://github.com/jhildenbiddle/css-vars-ponyfill/issues/158 for details.

2. Is this a library consumers may want to load from a CDN? If so, then I believe the library should offer minified as well as unminifed versions:

   ```html
   <!-- Minified UMD -->
   <script src="https://unpkg.com/balanced-match@3"></script>
   ```

   ```html
   <!-- Minified ESM -->
   <script type="module" src="https://unpkg.com/balanced-match@3/dist/balances-match.esm.min.js"></script>
   ```

Just my $0.02. :)

If you prefer going "ESM only", I'm happy to made those changes and update the docs accordingly. Given that previous versions of balanced-match were written in ES5, I think it's important to let consumers know that the current version requires transpilation if used in legacy environments.

Reply:
> For server-side libraries, sure. For projects that may run in a browser, it makes sense to continue offering an ES5 version a bit longer.

Interesting, in my experience it's the opposite. All of my client side projects are transpiled, none of my server side projects are, so I've been using ESM in the browser for a while. Is it the opposite for you?

If you were to consume and _not_ have a bundler set up, how would you do it? Is that a realistic use case for this library?

Reply:
I asked for community input in https://twitter.com/juliangruber/status/1391336252096516098

Reply:
> I asked for community input in https://twitter.com/juliangruber/status/1391336252096516098

The poll is interesting, but it misses the point of the PR. A better poll would be "should library authors pre-transpile distributable files to ES5 or are library consumers responsible for transpiling their dependencies to ES5?" I would expect the answer to be different for new libraries that have never offered ES5 versions versus well-established libraries that have always been ES5.

I think in 2021 there is a strong argument for moving to ES modules only, but the code in that module should still be ES5 *if* there is a reasonable chance it will be run in a legacy browser. Give the age and popularity of this library, I think that is a reasonable assumption.

Again, just my $0.02. :smile:

Reply:
i think we should just move to ESM